internal/dlog/calc.go

```diff
@@ -253,3 +325,102 @@ func (c *CalcBN256) BabyStepGiantStep(h, g *bn256.GT) (*big.Int, error) {
 
 	return nil, fmt.Errorf("failed to find discrete logarithm within bound")
 }
+
+// BabyStepGiantStep uses the baby-step giant-step method to
+// compute the discrete logarithm in the BN256.GT group. If c.neg is
+// set to true it searches for the answer within [-bound, bound].
+// It does so by running two goroutines, one for negative
+// answers and one for positive. If c.neg is set to false
+// only one goroutine is started, searching for the answer
+// within [0, bound].
+func (c *CalcBN256) BabyStepGiantStep(h, g *bn256.GT) (*big.Int, error) {
+	// create goroutines calculating positive and possibly negative
+	// result if c.neg is set to true
+	retChan := make(chan *big.Int)
+	errChan := make(chan error)
+	go c.runBabyStepGiantStepIterative(h, g, retChan, errChan)
+	if c.neg {
+		gInv := new(bn256.GT).Neg(g)
+		go c.runBabyStepGiantStepIterative(h, gInv, retChan, errChan)
+	}
+
+	// catch a value when the first routine finishes
+	ret := <-retChan
+	err := <-errChan
+
+	// prevent the situation when one routine exhausted all possibilities
+	// before the second found the solution
+	if c.neg && err != nil {
+		ret = <-retChan
+		err = <-errChan
+	}
+	// if both routines give an error, return an error
+	if err != nil {
+		return nil, err
+	}
+	// based on ret decide which routine gave the answer, thus if
+	// answer is negative
+	if c.neg && h.String() != new(bn256.GT).ScalarMult(g, ret).String() {
+		ret.Neg(ret)
+	}
+
+	return ret, err
+}
+
+// runBabyStepGiantStepIterative implements the baby-step giant-step method to
+// compute the discrete logarithm in the BN256.GT group. It is meant to be run
+// as a goroutine.
+//
+// The function searches for x, where h = g^x in BN256.GT group where operations
+// are written as multiplications. If the solution was not found
+// within the provided bound, it returns an error. In contrast to the usual
+// implementation of the method, this one proceeds iteratively, meaning that
+// smaller the solution is, faster the algorithm finishes.
+func (c *CalcBN256) runBabyStepGiantStepIterative(h, g *bn256.GT, retChan chan *big.Int, errChan chan error) {
+	one := big.NewInt(1)
+	two := big.NewInt(2)
+
+	// bn256.GT cannot be a key, thus we use a stringified bytes representation of the integer
```

Comment:
Nitpick: `bn256.GT` is not an integer, correct? `We use a stringified representation of the struct` would be right I think...

Reply:
Yes, you're right!